fendermint/vm/interpreter/src/chain.rs

```diff
@@ -420,22 +495,24 @@ fn relayed_bottom_up_ckpt_to_fvm(
     Ok(msg)
 }
 
-/// Encode to fvm implicit message
-pub fn encode_to_fvm_implicit(bytes: &[u8]) -> anyhow::Result<FvmMessage> {
-    let params = RawBytes::serialize(BytesSer(bytes))?;
-    let msg = FvmMessage {
-        version: 0,
-        from: system::SYSTEM_ACTOR_ADDR,
-        to: ipc::GATEWAY_ACTOR_ADDR,
-        value: TokenAmount::zero(),
-        method_num: ipc::gateway::METHOD_INVOKE_CONTRACT,
-        params,
-        // we are sending a implicit message, no need to set sequence
-        sequence: 0,
-        gas_limit: fvm_shared::BLOCK_GAS_LIMIT,
-        gas_fee_cap: TokenAmount::zero(),
-        gas_premium: TokenAmount::zero(),
-    };
+fn is_error(ret: &SignedMessageApplyRes) -> bool {
+    if let Ok(inner) = ret {
+        inner.fvm.apply_ret.msg_receipt.exit_code != ExitCode::OK
+    } else {
+        true
+    }
+}
 
-    Ok(msg)
+/// Derive the previous committed parent finality height
+pub fn derive_previous_height<DB: Blockstore>(
+    gateway: &GatewayCaller<DB>,
+    bytes: RawBytes,
+    provider: &TopDownFinalityProvider,
+) -> anyhow::Result<(u64, Option<IPCParentFinality>)> {
+    let (has_committed, prev_finality) = gateway.decode_commit_parent_finality_return(bytes)?;
+    Ok(if !has_committed {
+        (provider.genesis_epoch()?, None)
+    } else {
+        (prev_finality.height, Some(prev_finality))
+    })
```

Comment:
I think it would be good to move most of these into a `topdown` module, similar to the `checkpoint` module. They are so numerous and complicated, they completely dominate the `chain.rs` file by now.